README
NAME
    AnyEvent::XMPP - An implementation of the XMPP Protocol

VERSION
    Version 0.5

SYNOPSIS
       use AnyEvent::XMPP::Connection;

    or:

       use AnyEvent::XMPP::IM::Connection;

    or:

       use AnyEvent::XMPP::Client;

DESCRIPTION
    This is the head module of the AnyEvent::XMPP XMPP client protocol (as
    described in RFC 3920 and RFC 3921) framework.

    AnyEvent::XMPP::Connection is a RFC 3920 conforming "XML" stream
    implementation for clients, which handles TCP connect up to the resource
    binding. And provides low level access to the XML nodes on the XML
    stream along with some high level methods to send the predefined XML
    stanzas.

    AnyEvent::XMPP::IM::Connection is a more high level module, which is
    derived from AnyEvent::XMPP::Connection. It handles all the instant
    messaging client functionality described in RFC 3921.

    AnyEvent::XMPP::Client is a multi account client class. It manages
    connections to multiple XMPP accounts and tries to offer a nice high
    level interface to XMPP communication.

    For a list of "Supported extensions" see below.

    There are also other modules in this distribution, for example:
    AnyEvent::XMPP::Util, AnyEvent::XMPP::Writer, AnyEvent::XMPP::Parser and
    those I forgot :-) Those modules might be helpful and/or required if you
    want to use this framework for XMPP.

    See also AnyEvent::XMPP::Writer for a discussion about the brokenness of
    XML in the XMPP specification.

    If you have any questions or seek for help look below under "SUPPORT".

REQUIREMENTS
    One of the major drawbacks I see for AnyEvent::XMPP is the long list of
    required modules to make it work.

    AnyEvent
        For the I/O events, timers, TCP, TLS, DNS and I/O buffering.

    Object::Event
        The former AnyEvent::XMPP::Event module has been outsourced to the
        Object::Event module to provide a more generic way for more other
        modules to register and call event callbacks.

    XML::Writer
        For writing "XML".

    XML::Parser::Expat
        For parsing partial "XML" stuff.

    MIME::Base64
        For SASL authentication

    Authen::SASL
        For SASL authentication

    Net::LibIDN
        For stringprep profiles to handle JIDs.

    Digest::SHA1
        For component authentication and old-style authentication.

    And yes, all these are essential for XMPP communication. Even though
    'instant messaging' and 'presence' is a quite simple problem XMPP
    somehow was successful at making the task complicated enough to keep me
    busy for a long time. But all of that time wasn't only for the
    technology required to get it started, mostly it was for all the quirks,
    hacks and badly applied "XML" in the protocol which complicated the
    matter.

RELEASE NOTES
    Here are some notes to the last releases (release of this version is at
    top):

  Version
    *   0.5

        Maintenance release. Added a patch from Marcus Dubois for
        Ext::Pubsub. Also fixed some memleaks in AnyEvent::XMPP::Parser.

        Also wanted to note that the next version of AnyEvent::XMPP will
        have an incompatible API. If you are eager to try out the new
        complete rewrite of AnyEvent::XMPP contact me.

    *   0.4

        Minor fixes and feature enhancements: Added old_style_ssl option for
        direct port 5223 SSL connections. Providing 'get_own_contact' for
        keeping track of own resources.

        The AnyEvent::XMPP::Ext::MUC extension was rewritten and provides a
        more sane API now.

        For details consult the Changes file in the distribution.

    *   0.3

        Fixed some small bugs and improved documentation a bit, especially
        w.r.t. parameter passing of host and ports.

    *   0.2

        Renamed module from Net::XMPP2 to AnyEvent::XMPP. Net::XMPP2 is
        herby deprecated!

        Rewrote the low-level socket stuff to use AnyEvent::Socket and
        AnyEvent::Handle. Removed blocking write functionality, which can't
        be supported that easily with AnyEvent::Handle (however, if you want
        to wait until the send-buffer is empty you best use the
        "send_buffer_empty" event of AnyEvent::XMPP::Connection).

        For more details consult the Changes file of the AnyEvent::XMPP
        distribution.

    *   older

        For older release notes please have a look at the Changes file or
        CPAN.

  TODO
    There are still lots of items on the TODO list (see also the TODO file
    in the distribution of AnyEvent::XMPP).

TEST SUITE
    If you are a developer and want to test either a server or maybe just
    whether this module passes some basic tests you might want to run the
    developer test suite.

    This test suite is not enabled by default because it requires some human
    interaction to set it up, please see AnyEvent::XMPP::TestClient for
    hints about the setup procedure for the test suite.

    I wrote the test suite mostly because I wanted to make sure I didn't
    break something essential before a release. The tests don't cover
    everything and I don't plan to write a test for every single function in
    the API, that would slow down development considerably for me. But I
    hope that some grave show stopper bugs in releases are prevented with
    this test suite.

    The tests are also useful if you want to test a server implementation.
    But there are maybe of course conformance issues with AnyEvent::XMPP
    itself, so if you find something where AnyEvent::XMPP doesn't conform to
    the XMPP RFCs or XEPs consult the BUGS section below.

    If you find a server that doesn't handle something correctly but you
    need to interact with it you are free to implement workarounds and send
    me a patch, or even ask me whether I might want to look into the issue
    (I can't guarantee anything here, but I want this module to be as
    interoperable as possible. But if the implementation of a workaround for
    some non-conformant software will complicate the code too much I'm
    probably not going to implement it.).

    Of course, if you find a bug in some server implementation don't forget
    to file a bugreport to them, one hack less in AnyEvent::XMPP means more
    time for bug fixing and improvements and new features.

Why (yet) another XMPP module?
    The main outstanding feature of this module in comparison to the other
    XMPP (aka Jabber) modules out there is the support for AnyEvent.
    AnyEvent permits you to use this module together with other I/O event
    based programs and libraries (ie. Gtk2 or Event).

    The other modules could often only be integrated in those applications
    or libraries by using threads. I decided to write this module because I
    think CPAN lacks an event based XMPP module. Threads are unfortunately
    not an alternative in Perl at the moment due the limited threading
    functionality they provide and the global speed hit. I also think that a
    simple event based I/O framework might be a bit easier to handle than
    threads.

    Another thing was that I didn't like the APIs of the other modules. In
    AnyEvent::XMPP I try to provide low level modules for speaking XMPP as
    defined in RFC 3920 and RFC 3921 (see also AnyEvent::XMPP::Connection
    and AnyEvent::XMPP::IM::Connection). But I also try to provide a high
    level API for easier usage for instant messaging tasks and clients (eg.
    AnyEvent::XMPP::Client).

Supported extensions
    See AnyEvent::XMPP::Ext for a list.

EXAMPLES
    Following examples are included in this distribution:

    samples/simple_example_1
        This example script just connects to a server and sends a message
        and also displays incoming messages on stdout.

    samples/conference_lister
        See below.

    samples/room_lister
        See below.

    samples/room_lister_stat
        These three scripts implements a global room scan.
        "conference_lister" takes a list of servers (the file is called
        "servers.xml" which has the same format as the xml file at
        <http://www.jabber.org/servers.xml>). It then scans all servers for
        chat room services and lists them into a file "conferences.stor",
        which is a Storable dump.

        "room_lister" then reads that file and queries all services for
        rooms, and then all rooms for their occupants. The output file is
        "room_data.stor", also a Storable dump, which in turn can be read
        with "room_lister_stat", which transform the data structures into
        something human readable.

        These scripts are a bit hacky and quite complicated, but maybe it's
        of any value for someone. You might note "EVQ.pm" in samples which
        is a module that handles request-throttling (You don't want to flood
        the server and risk getting the admins attention :).

    samples/simple_component
        This is a (basic) skeleton for a jabber component.

    samples/simple_oob_retriever
        This is a simple out of band file transfer receiver bot. It uses
        "curl" to fetch the files and also has the sample functionality of
        sending a file url for someone who sends the bot a 'send <filename>'
        message.

    samples/simple_register_example
        This is a example script which allows you to register, unregister
        and change your password for accounts. Execute it without arguments

Changes
Revision history for AnyEvent::XMPP:

         - small uncleanliness fixed in ::Parser.

0.5 Mon Mar 16 21:13:13 CET 2009

   !!!!!!!!!!!!!!!
   !!! NOTE: The next version of AnyEvent::XMPP will have an incompatible API!
   !!!!!!!!!!!!!!!

         - added patch for Ext::Pubsub.
         - melo fixed some memleaks.
         - fixed another long time memleak in the parser.

0.4 Sun Feb 15 22:59:38 CET 2009
         - added a $dont_retrieve_vcard to ::Ext::VCard, to
           prevent retrieval when session_ready event comes.
         - implemented support for old-style (tcp port 5223) SSL connections,
           use the old_style_ssl option for ::Connection to enable TLS handshaking
           upon TCP connect.
         - added $message argument to subscription related events. the contents
           comes from the <status> elements in the presence stanza.
         - made special contact for keeping track of the presences of our own
           resources. See ::IM::Roster::get_own_contact() and ::IM::Contact::is_me().
         - minor fixes in ::Ext::MUC.
         - added 'stream_version_override' parameter to ::Connection, mostly
           for testing purposes with ejabberd (to enforce old-jabber-style
           authentication).
         - partially reversed a patch w.r.t. hostname usage for SASL mechanism.
           A flag 'use_host_as_sasl_hostname' is provided for giving the hostname
           to SASL.
         - rewrote AnyEvent::XMPP::Ext::MUC a bit and made the API more sane.
           for example: join_room doesn't take a callback anymore.
         - added more C<disco_feature> methods to the extensions, so that
           they can be registered to an ::Ext::Disco extension to display
           support for certain features.
         - xmpp_datetime_as_timestamp fixed w.r.t. UTC based timestamps.
         - added possibility to override iq_xml events, and documented
           how to overtake handling of message, presence and iq stanzas.
         - added patch from Pedro Melo to give stanzas sent by a ::Component
           the right namespace.
         - added patch from Pedro Melo to fix passing of the host in ::Component.

0.3 Mon Jan 28 23:00:58 CET 2009
         - fixed a bug in disconnect error reporting.
         - fixed bad parameter passing of AnyEvent::XMPP::Client to
           AnyEvent::XMPP::IM::Account, also 'host' is used for overriding
           the destination host now (and not 'domain', as it was in version
           0.2).
         - half fixed RT #42194 (Domain vs. host and GSSAPI),
           passing the hostname specified by 'host' for ::Connection to
           Authen::SASL's client_new method.
         - fixed bug in ::XMPP::Client where remove_account didn't work
           as expected (and set_accounts).

0.2 Sun Dec 28 15:09:25 CET 2008
         - Using the better tested AnyEvent::Handle and
           AnyEvent::Socket for DNS resolution, TCP connect and TLS now.
         - override_host and override_port have been removed.
         - weird DNS SRV handling and comments have been removed.
         - removed blocked_write option and drain function, send_buffer_empty event
           should be used. blocking behaviour should be emulated with send_buffer_empty
           at the toplevel of the program.
         - improved message tracking api in ::IM::Account
         - added added_account/removed_account events to ::Client
         - added set_accounts() function to ::Client
         - removed unneccessary sample (devcl)
         - renamed module from Net::XMPP2 to AnyEvent::XMPP
         - minor additions to ::Ext::Pubsub.

0.14 Fri Aug  1 12:13:11 CEST 2008
         - NOTICE: This is the last release of this module under the name
                   Net::XMPP2. All further releases will be done under the name
                   AnyEvent::XMPP!
         - applied fixes from git://github.com/melo/net--xmpp2.git
           branch component-reply-with-from.
         - fixed small bug in simple_component.
         - fixed a debugging bug in the drain method of SimpleConnection,
           where the written data wasn't passed to the debugging events.
         - improved the documentation a bit and fixed typos.

0.13 Tue Apr 15 13:07:02 CEST 2008
         - added a 'send_buffer_empty' event, to be able to wait
           for an empty send buffer. and also the C<drain> method, which
           will block until the send buffer is empty. also added a C<blocking_write>
           option for L<Net::XMPP2::Connection>.
         - added a default handler to L<Net::XMPP2::Parser> which
           recognizes and stores CDATA now correctly, so that it can

lib/Net/XMPP2/Writer.pm
package Net::XMPP2::Writer;
use strict;
use XML::Writer;
use Authen::SASL qw/Perl/;
use MIME::Base64;
use Net::XMPP2::Namespaces qw/xmpp_ns/;
use Net::XMPP2::Util qw/simxml filter_xml_chars filter_xml_attr_hash_chars/;
use Digest::SHA1 qw/sha1_hex/;
use Encode;

=head1 NAME

Net::XMPP2::Writer - "XML" writer for XMPP

=head1 SYNOPSIS

   use Net::XMPP2::Writer;
   ...

=head1 DESCRIPTION

This module contains some helper functions for writing XMPP "XML",
which is not real XML at all ;-( I use L<XML::Writer> and tune it
until it creates "XML" that is accepted by most servers propably
(all of the XMPP servers I tested should work (jabberd14, jabberd2,
ejabberd, googletalk).

I hope the semantics of L<XML::Writer> don't change much over the future,
but if they do and you run into problems, please report them!

The whole "XML" concept of XMPP is fundamentally broken anyway. It's supposed
to be an subset of XML. But a subset of XML productions is not XML. Strictly
speaking you need a special XMPP "XML" parser and writer to be 100% conformant.

On top of that XMPP B<requires> you to parse these partial "XML" documents.
But a partial XML document is not well-formed, heck, it's not even a XML document!.
And a parser should bail out with an error. But XMPP doesn't care, it just relies on
implementation dependend behaviour of chunked parsing modes for SAX parsing.
This functionality isn't even specified by the XML recommendation in any way.
The recommendation even says that it's undefined what happens if you process
not-well-formed XML documents.

But I try to be as "XML" XMPP conformant as possible (it should be around 99-100%).
But it's hard to say what XML is conformant, as the specifications of XMPP "XML" and XML
are contradicting. For example XMPP also says you only have to generated and accept
utf-8 encodings of XML, but the XML recommendation says that each parser has
to accept utf-8 B<and> utf-16. So, what do you do? Do you use a XML conformant parser
or do you write your own?

I'm using XML::Parser::Expat because expat knows how to parse broken (aka 'partial')
"XML" documents, as XMPP requires. Another argument is that if you capture a XMPP
conversation to the end, and even if a '</stream:stream>' tag was captured, you
wont have a valid XML document. The problem is that you have to resent a <stream> tag
after TLS and SASL authentication each! Awww... I'm repeating myself.

But well... Net::XMPP2 does it's best with expat to cope with the fundamental brokeness
of "XML" in XMPP.

Back to the issue with "XML" generation: I've discoverd that many XMPP servers (eg.
jabberd14 and ejabberd) have problems with XML namespaces. Thats the reason why
I'm assigning the namespace prefixes manually: The servers just don't accept validly
namespaced XML. The draft 3921bis does even state that a client SHOULD generate a 'stream'
prefix for the <stream> tag.

I advice you to explictly set the namespaces too if you generate "XML" for XMPP yourself,
at least until all or most of the XMPP servers have been fixed. Which might take some
years :-) And maybe will happen never.

And another note: As XMPP requires all predefined entity characters to be escaped
in character data you need a "XML" writer that will escape everything:

   RFC 3920 - 11.1.  Restrictions:

     character data or attribute values containing unescaped characters
     that map to the predefined entities (Section 4.6 therein);
     such characters MUST be escaped

This means:
You have to escape '>' in the character data. I don't know whether XML::Writer
does that. And I honestly don't care much about this. XMPP is broken by design and
I have barely time to writer my own XML parsers and writers to suit their sick taste
of "XML". (Do I repeat myself?)

I would be happy if they finally say (in RFC3920): "XMPP is NOT XML. It's just
XML-like, and some XML utilities allow you to process this kind of XML.".

=head1 METHODS

=over 4

=item B<new (%args)>

This methods takes following arguments:

=over 4

=item write_cb

The callback that is called when a XML stanza was completly written
and is ready for transfer. The first argument of the callback
will be the character data to send to the socket.

=back

And calls C<init>.

=cut

sub new {
   my $this = shift;
   my $class = ref($this) || $this;
   my $self = {
      write_cb     => sub {},
      send_iq_cb   => sub {},
      send_msg_cb  => sub {},
      send_pres_cb => sub {},
      @_
   };
   bless $self, $class;
   $self->init;
   return $self;
}

=item B<init>

(Re)initializes the writer.

=cut

sub init {
   my ($self) = @_;
   $self->{write_buf} = "";
   $self->{writer} =
      XML::Writer->new (OUTPUT => \$self->{write_buf}, NAMESPACES => 1, UNSAFE => 1);
}

=item B<flush ()>

This method flushes the internal write buffer and will invoke the C<write_cb>
callback. (see also C<new ()> above)

=cut

sub flush {
   my ($self) = @_;
   $self->{write_cb}->(substr $self->{write_buf}, 0, (length $self->{write_buf}), '');
}

=item B<send_init_stream ($language, $domain, $namespace)>

This method will generate a XMPP stream header. C<$domain> has to be the
domain of the server (or endpoint) we want to connect to.

C<$namespace> is the namespace uri or the tag (from L<Net::XMPP2::Namespaces>)
for the stream namespace. (This is used by L<Net::XMPP2::Component> to connect
as component to a server). C<$namespace> can also be undefined, in this case
the C<client> namespace will be used.

=cut

sub send_init_stream {
   my ($self, $language, $domain, $ns) = @_;

   $ns ||= 'client';

   my $w = $self->{writer};
   $w->xmlDecl ();
   $w->addPrefix (xmpp_ns ('stream'), 'stream');
   $w->addPrefix (xmpp_ns ($ns), '');
   $w->forceNSDecl (xmpp_ns ($ns));
   $w->startTag (
      [xmpp_ns ('stream'), 'stream'],
      to      => $domain,
      version => '1.0',
      [xmpp_ns ('xml'), 'lang'] => $language
   );
   $self->flush;
}

=item B<send_whitespace_ping>

This method sends a single space to the server.

=cut

sub send_whitespace_ping {
   my ($self) = @_;
   $self->{writer}->raw (' ');
   $self->flush;
}

=item B<send_handshake ($streamid, $secret)>

This method sends a component handshake. Please note that C<$secret>
must be XML escaped!

=cut

sub send_handshake {
   my ($self, $id, $secret) = @_;
   my $out_secret = encode ("UTF-8", $secret);
   my $out = lc sha1_hex ($id . $out_secret);
   simxml ($self->{writer}, defns => 'component', node => {
      ns => 'component', name => 'handshake', childs => [ $out ]
   });
   $self->flush;
}

=item B<send_end_of_stream>

Sends end of the stream.

=cut

sub send_end_of_stream {
   my ($self) = @_;
   my $w = $self->{writer};
   $w->endTag ([xmpp_ns ('stream'), 'stream']);
   $self->flush;
}

=item B<send_sasl_auth ($mechanisms)>

This methods sends the start of a SASL authentication. C<$mechanisms> is
a string with space seperated mechanisms that are supported by the other
end.

=cut

sub send_sasl_auth {
   my ($self, $mechanisms, $user, $domain, $pass) = @_;

   my $sasl = Authen::SASL->new (
      mechanism => $mechanisms,
      callback => {
 # XXX: removed authname, because it ensures maximum connectivitiy
 #      along multiple server implementations - XMPP is such a crap
 #        authname => $user . '@' . $domain,
         user => $user,
         pass => $pass,
      }
   );

   $self->{sasl} = $sasl->client_new ('xmpp', $domain);

   my $w = $self->{writer};
   $w->addPrefix (xmpp_ns ('sasl'),   '');
   $w->startTag ([xmpp_ns ('sasl'), 'auth'], mechanism => $self->{sasl}->mechanism);
   $w->characters (MIME::Base64::encode_base64 ($self->{sasl}->client_start, ''));
   $w->endTag;
   $self->flush;
}

=item B<send_sasl_response ($challenge)>

This method generated the SASL authentication response to a C<$challenge>.
You must not call this method without calling C<send_sasl_auth ()> before.

=cut

sub send_sasl_response {
   my ($self, $challenge) = @_;
   $challenge = MIME::Base64::decode_base64 ($challenge);
   my $ret = '';
   unless ($challenge =~ /rspauth=/) { # rspauth basically means: we are done
      $ret = $self->{sasl}->client_step ($challenge);
      unless ($ret) {
         die "Error in SASL authentication in client step with challenge: '$challenge'\n";
      }
   }
   my $w = $self->{writer};
   $w->addPrefix (xmpp_ns ('sasl'),   '');
   $w->startTag ([xmpp_ns ('sasl'), 'response']);
   $w->characters (MIME::Base64::encode_base64 ($ret, ''));
   $w->endTag;
   $self->flush;
}

=item B<send_starttls>

Sends the starttls command to the server.

=cut

sub send_starttls {
   my ($self) = @_;
   my $w = $self->{writer};
   $w->addPrefix (xmpp_ns ('tls'),   '');
   $w->emptyTag ([xmpp_ns ('tls'), 'starttls']);
   $self->flush;
}

=item B<send_iq ($id, $type, $create_cb, %attrs)>

This method sends an IQ stanza of type C<$type> (to be compliant
only use: 'get', 'set', 'result' and 'error').

If C<$create_cb> is a code reference it will be called with an XML::Writer
instance as first argument, which must be used to fill the IQ stanza. The
XML::Writer is in UNSAFE mode, so you can safely use C<raw()> to write out XML.

C<$create_cb> is a hash reference the hash will be used as key=>value arguments
for the C<simxml> function defined in L<Net::XMPP2::Util>. C<simxml> will then
be used to generate the contents of the IQ stanza. (This is very convenient
when you want to write the contents of stanzas in the code and don't want to
build a DOM tree yourself...).

If C<$create_cb> is an array reference it's elements will be interpreted as
single C<$create_cb> argument (which can either be a hash reference or code
reference themself) and executed sequencially.

If C<$create_cb> is undefined an empty tag will be generated.

Example:

   $writer->send_iq ('newid', 'get', {
      defns => 'version',
      node  => { name => 'query', ns => 'version' }
   }, to => 'jabber.org')

C<%attrs> should have further attributes for the IQ stanza tag.
For example 'to' or 'from'. If the C<%attrs> contain a 'lang' attribute
it will be put into the 'xml' namespace. If the 'to' attribute contains
an undef it will be omitted.

C<$id> is the id to give this IQ stanza and is mandatory in this API.

Please note that all attribute values and character data will be filtered
by C<filter_xml_chars> (see also L<Net::XMPP2::Util>).

=cut

sub send_iq {
   my ($self, $id, $type, $create_cb, %attrs) = @_;

   $create_cb = _trans_create_cb ($create_cb);
   $create_cb = $self->_fetch_cb_additions (send_iq_cb => $create_cb, $id, $type, \%attrs);

   my (@from) = ($self->{jid} ? (from => $self->{jid}) : ());
   if ($attrs{lang}) {
      push @from, ([ xmpp_ns ('xml'), 'lang' ] => delete $attrs{leng})
   }

   unless (defined $attrs{to}) {
      delete $attrs{to};
   }

   push @from, (id => filter_xml_chars $id) if defined $id;

   filter_xml_attr_hash_chars \%attrs;

   my $w = $self->{writer};
   $w->addPrefix (xmpp_ns ('client'), '');

   if (defined $create_cb) {
      $w->startTag ([xmpp_ns ('client'), 'iq'], type => $type, @from, %attrs);
      $create_cb->($w);
      $w->endTag;
   } else {
      $w->emptyTag ([xmpp_ns ('client'), 'iq'], type => $type, @from, %attrs);
   }
   $self->flush;
}

=item B<send_presence ($id, $type, $create_cb, %attrs)>

Sends a presence stanza.

C<$create_cb> has the same meaning as for C<send_iq>.
C<%attrs> will let you pass further optional arguments like 'to'.

C<$type> is the type of the presence, which may be one of:

   unavailable, subscribe, subscribed, unsubscribe, unsubscribed, probe, error

Or undef, in case you want to send a 'normal' presence.
Or something completly different if you don't like the RFC 3921 :-)

C<%attrs> contains further attributes for the presence tag or may contain one of the
following exceptional keys:

If C<%attrs> contains a 'show' key: a child xml tag with that name will be geenerated
with the value as the content, which should be one of 'away', 'chat', 'dnd' and 'xa'.

If C<%attrs> contains a 'status' key: a child xml tag with that name will be generated
with the value as content. If the value of the 'status' key is an hash reference
the keys will be interpreted as language identifiers for the xml:lang attribute
of each status element. If one of these keys is the empty string '' no xml:lang attribute
will be generated for it. The values will be the character content of the status tags.

If C<%attrs> contains a 'priority' key: a child xml tag with that name will be generated
with the value as content, which must be a number between -128 and +127.

Note: If C<$create_cb> is undefined and one of the above attributes (show,
status or priority) were given, the generates presence tag won't be empty.

Please note that all attribute values and character data will be filtered
by C<filter_xml_chars> (see also L<Net::XMPP2::Util>).

=cut

sub _generate_key_xml {
   my ($w, $key, $value) = @_;
   $w->startTag ($key);
   $w->characters (filter_xml_chars $value);
   $w->endTag;
}

sub _generate_key_xmls {
   my ($w, $key, $value) = @_;
   if (ref ($value) eq 'HASH') {
      for (keys %$value) {
         $w->startTag ($key, ($_ ne '' ? ([xmpp_ns ('xml'), 'lang'] => $_) : ()));
         $w->characters (filter_xml_chars $value->{$_});
         $w->endTag;
      }
   } else {
      $w->startTag ($key);
      $w->characters (filter_xml_chars $value);
      $w->endTag;
   }
}

sub _trans_create_cb {
   my ($cb) = @_;
   return unless defined $cb;
   if (ref ($cb) eq 'HASH') {
      my $args = $cb;
      $cb = sub {
         my ($w) = @_;
         simxml ($w, %$args);
      }
   } elsif (ref ($cb) eq 'ARRAY') {
      my @cbs = map { _trans_create_cb ($_) } @$cb;
      $cb = sub {
         my ($w) = @_;
         for (@cbs) { $_->($w) }
      }
   }
   $cb
}

sub _fetch_cb_additions {
   my ($self, $key, $create_cb, @args) = @_;
   my (@add_cbs) = $self->{$key}->(@args);
   @add_cbs = map { _trans_create_cb ($_) } @add_cbs;

   if (@add_cbs) {
      my $crcb = $create_cb;
      $create_cb = sub {
         my (@args) = @_;
         $crcb->(@args) if $crcb;
         for (@add_cbs) { $_->(@args) }
      }
   }

   $create_cb
}

sub send_presence {
   my ($self, $id, $type, $create_cb, %attrs) = @_;

   $create_cb = _trans_create_cb ($create_cb);
   $create_cb = $self->_fetch_cb_additions (send_pres_cb => $create_cb, $id, $type, \%attrs);

   my $w = $self->{writer};
   $w->addPrefix (xmpp_ns ('client'), '');

   my @add;
   push @add, (type => $type) if defined $type;
   push @add, (id => $id) if defined $id;

   my %fattrs =
      map { $_ => $attrs{$_} }
         grep { my $k = $_; not grep { $k eq $_ } qw/show priority status/ }
            keys %attrs;

   filter_xml_attr_hash_chars \%fattrs;

   if (defined $create_cb) {
      $w->startTag ([xmpp_ns ('client'), 'presence'], @add, %fattrs);
      _generate_key_xml ($w, show => $attrs{show})         if defined $attrs{show};
      _generate_key_xml ($w, priority => $attrs{priority}) if defined $attrs{priority};
      _generate_key_xmls ($w, status => $attrs{status})    if defined $attrs{status};
      $create_cb->($w);
      $w->endTag;
   } else {
      if (exists $attrs{show} or $attrs{priority} or $attrs{status}) {
         $w->startTag ([xmpp_ns ('client'), 'presence'], @add, %fattrs);
         _generate_key_xml ($w, show => $attrs{show})         if defined $attrs{show};
         _generate_key_xml ($w, priority => $attrs{priority}) if defined $attrs{priority};
         _generate_key_xmls ($w, status => $attrs{status})    if defined $attrs{status};
         $w->endTag;
      } else {
         $w->emptyTag ([xmpp_ns ('client'), 'presence'], @add, %fattrs);
      }
   }

   $self->flush;
}

=item B<send_message ($id, $to, $type, $create_cb, %attrs)>

Sends a message stanza.

C<$to> is the destination JID of the message. C<$type> is
the type of the message, and if C<$type> is undefined it will default to 'chat'.
C<$type> must be one of the following: 'chat', 'error', 'groupchat', 'headline'
or 'normal'.

C<$create_cb> has the same meaning as in C<send_iq>.

C<%attrs> contains further attributes for the message tag or may contain one of the
following exceptional keys:

If C<%attrs> contains a 'body' key: a child xml tag with that name will be generated
with the value as content. If the value of the 'body' key is an hash reference
the keys will be interpreted as language identifiers for the xml:lang attribute
of each body element. If one of these keys is the empty string '' no xml:lang attribute
will be generated for it. The values will be the character content of the body tags.

If C<%attrs> contains a 'subject' key: a child xml tag with that name will be generated
with the value as content. If the value of the 'subject' key is an hash reference
the keys will be interpreted as language identifiers for the xml:lang attribute
of each subject element. If one of these keys is the empty string '' no xml:lang attribute
will be generated for it. The values will be the character content of the subject tags.

If C<%attrs> contains a 'thread' key: a child xml tag with that name will be generated
and the value will be the character content.

Please note that all attribute values and character data will be filtered
by C<filter_xml_chars> (see also L<Net::XMPP2::Util>).

=cut

sub send_message {
   my ($self, $id, $to, $type, $create_cb, %attrs) = @_;

   $create_cb = _trans_create_cb ($create_cb);
   $create_cb = $self->_fetch_cb_additions (send_msg_cb => $create_cb, $id, $to, $type, \%attrs);

   my $w = $self->{writer};
   $w->addPrefix (xmpp_ns ('client'), '');

   my @add;
   push @add, (id => $id) if defined $id;

   $type ||= 'chat';

   my %fattrs =
      map { $_ => $attrs{$_} }
         grep { my $k = $_; not grep { $k eq $_ } qw/subject body thread/ }
            keys %attrs;

   if (defined $create_cb) {
      $w->startTag ([xmpp_ns ('client'), 'message'], @add, to => $to, type => $type, %fattrs);
      _generate_key_xmls ($w, subject => $attrs{subject})    if defined $attrs{subject};
      _generate_key_xmls ($w, body => $attrs{body})          if defined $attrs{body};
      _generate_key_xml ($w, thread => $attrs{thread})       if defined $attrs{thread};
      $create_cb->($w);
      $w->endTag;
   } else {
      if (exists $attrs{subject} or $attrs{body} or $attrs{thread}) {
         $w->startTag ([xmpp_ns ('client'), 'message'], @add, to => $to, type => $type, %fattrs);
         _generate_key_xmls ($w, subject => $attrs{subject})    if defined $attrs{subject};
         _generate_key_xmls ($w, body => $attrs{body})          if defined $attrs{body};
         _generate_key_xml ($w, thread => $attrs{thread})       if defined $attrs{thread};
         $w->endTag;
      } else {
         $w->emptyTag ([xmpp_ns ('client'), 'message'], @add, to => $to, type => $type, %fattrs);
      }
   }

   $self->flush;
}


=item B<write_error_tag ($error_stanza_node, $error_type, $error)>

C<$error_type> is one of 'cancel', 'continue', 'modify', 'auth' and 'wait'.
C<$error> is the name of the error tag child element. If C<$error> is one of
the following:

   'bad-request', 'conflict', 'feature-not-implemented', 'forbidden', 'gone',
   'internal-server-error', 'item-not-found', 'jid-malformed', 'not-acceptable',
   'not-allowed', 'not-authorized', 'payment-required', 'recipient-unavailable',
   'redirect', 'registration-required', 'remote-server-not-found',
   'remote-server-timeout', 'resource-constraint', 'service-unavailable',
   'subscription-required', 'undefined-condition', 'unexpected-request'

then a default can be select for C<$error_type>, and the argument can be undefined.

Note: This method is currently a bit limited in the generation of the xml
for the errors, if you need more please contact me.

=cut

our %STANZA_ERRORS = (
   'bad-request'             => ['modify', 400],
   'conflict'                => ['cancel', 409],
   'feature-not-implemented' => ['cancel', 501],
   'forbidden'               => ['auth',   403],
   'gone'                    => ['modify', 302],
   'internal-server-error'   => ['wait',   500],
   'item-not-found'          => ['cancel', 404],
   'jid-malformed'           => ['modify', 400],
   'not-acceptable'          => ['modify', 406],
   'not-allowed'             => ['cancel', 405],
   'not-authorized'          => ['auth',   401],
   'payment-required'        => ['auth',   402],
   'recipient-unavailable'   => ['wait',   404],
   'redirect'                => ['modify', 302],
   'registration-required'   => ['auth',   407],
   'remote-server-not-found' => ['cancel', 404],
   'remote-server-timeout'   => ['wait',   504],
   'resource-constraint'     => ['wait',   500],
   'service-unavailable'     => ['cancel', 503],
   'subscription-required'   => ['auth',   407],
   'undefined-condition'     => ['cancel', 500],
   'unexpected-request'      => ['wait',   400],
);

sub write_error_tag {
   my ($self, $errstanza, $type, $error) = @_;

   my $w = $self->{writer};

   $_->write_on ($w) for $errstanza->nodes;

   my @add;

   unless (defined $type and defined $STANZA_ERRORS{$error}) {
      $type = $STANZA_ERRORS{$error}->[0];
   }

   push @add, (code => $STANZA_ERRORS{$error}->[1]);

   my %add = @add;
   filter_xml_attr_hash_chars \%add;

   $w->addPrefix (xmpp_ns ('client'), '');
   $w->startTag ([xmpp_ns ('client') => 'error'], type => $type, %add);
      $w->addPrefix (xmpp_ns ('stanzas'), '');
      $w->emptyTag ([xmpp_ns ('stanzas') => filter_xml_chars $error]);
   $w->endTag;
}

=back

=head1 AUTHOR

Robin Redeker, C<< <elmex at ta-sa.org> >>, JID: C<< <elmex at jabber.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2007 Robin Redeker, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Net::XMPP2
